quickwit/quickwit-storage/src/cache/byte_range_cache.rs

```diff
@@ -0,0 +1,359 @@
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::borrow::{Borrow, Cow};
+use std::collections::BTreeMap;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::Mutex;
+
+use tantivy::directory::OwnedBytes;
+
+use crate::metrics::CacheMetrics;
+
+#[derive(Clone, PartialOrd, Ord, PartialEq, Eq)]
+struct CacheKey<'a, T: ToOwned + ?Sized> {
+    tag: Cow<'a, T>,
+    range_start: usize,
+}
+
+impl<T: ToOwned + ?Sized> CacheKey<'static, T> {
+    fn from_owned(tag: T::Owned, range_start: usize) -> Self {
+        CacheKey {
+            tag: Cow::Owned(tag),
+            range_start,
+        }
+    }
+}
+
+impl<'a, T: ToOwned + ?Sized> CacheKey<'a, T> {
+    fn from_borrowed(tag: &'a T, range_start: usize) -> Self {
+        CacheKey {
+            tag: Cow::Borrowed(tag),
+            range_start,
+        }
+    }
+}
+
+struct CacheValue {
+    range_end: usize,
+    bytes: OwnedBytes,
+}
+
+// T is a tag, usually a file path, R is the range bound type
+struct NeedMutByteRangeCache<T: 'static + ToOwned + ?Sized> {
+    cache: BTreeMap<CacheKey<'static, T>, CacheValue>,
+    // this is hardly significant as items can get merged if they overlap
+    num_items: u64,
+    num_bytes: u64,
+    cache_counters: &'static CacheMetrics,
+}
+
+impl<T: 'static + ToOwned + ?Sized + Ord> NeedMutByteRangeCache<T> {
+    fn with_infinite_capacity(cache_counters: &'static CacheMetrics) -> Self {
+        NeedMutByteRangeCache {
+            cache: BTreeMap::new(),
+            num_items: 0,
+            num_bytes: 0,
+            cache_counters,
+        }
+    }
+
+    fn get_slice(&mut self, tag: &T, byte_range: Range<usize>) -> Option<OwnedBytes> {
+        if byte_range.start == byte_range.end {
+            return Some(OwnedBytes::empty());
+        }
+
+        let key = CacheKey::from_borrowed(tag, byte_range.start);
+        let (k, v) = if let Some((k, v)) = self.get_block(&key, byte_range.end) {
+            (k, v)
+        } else {
+            self.cache_counters.misses_num_items.inc();
+            return None;
+        };
+
+        let start = byte_range.start - k.range_start;
+        let end = byte_range.end - k.range_start;
+
+        self.cache_counters.hits_num_items.inc();
+        self.cache_counters
+            .hits_num_bytes
+            .inc_by((end - start) as u64);
+
+        Some(v.bytes.slice(start..end))
+    }
+
+    fn put_slice(&mut self, tag: T::Owned, byte_range: Range<usize>, bytes: OwnedBytes) {
+        let len = byte_range.end - byte_range.start;
```

Comment:
I suspect this code could have been mildly simpler... but it is well encapsulated and well tested, so actually I think we don't care.